### PR TITLE
feat: Builder pattern for VirtualMachine

### PR DESCRIPTION
**Pull request by hurrikaanig**
# Builder pattern for VirtualMachine

## Description

Added a builder pattern for VirtualMachine (VirtualMachineBuilder) to allow creating VirtualMachine with customized fields.

Linked to issue #819 

## Checklist
- [X] Linked to Github Issue
- [X] Unit tests added
- [ ] Integration tests added.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.

### CODE REVIEW

File: src/vm/vm_core.rs [hunk 1]
```diff
@@ -996,6 +996,138 @@ impl VirtualMachine {
...
+    pub fn builtin_runners(
+        mut self,
+        builtin_runners: Vec<(String, BuiltinRunner)>,
+    ) -> VirtualMachineBuilder {
+        self.builtin_runners = builtin_runners;
+        self
+    }
+
+    pub fn segments(mut self, segments: MemorySegmentManager) -> VirtualMachineBuilder {
+        // Set the name on the builder itself, and return the builder by value.
```
Comment (tdelabro) [inline src/vm/vm_core.rs:1054]: You copy pasted this comment on every setter method.
Yon can remove it every time

File: src/vm/vm_core.rs [hunk 2]
```diff
@@ -996,6 +996,138 @@ impl VirtualMachine {
...
+
+    pub fn run_finished(mut self, run_finished: bool) -> VirtualMachineBuilder {
+        // Set the name on the builder itself, and return the builder by value.
+        self.run_finished = run_finished;
+        self
+    }
+
+    // If we can get away with not consuming the Builder here, that is an
+    // advantage. It means we can use the FooBuilder as a template for constructing
+    // many Foos.
```
Comment (tdelabro) [inline src/vm/vm_core.rs:1112]: Remove this comment

File: src/vm/vm_core.rs [hunk 3]
```diff
@@ -996,6 +996,138 @@ impl VirtualMachine {
...
+    pub fn build(self) -> VirtualMachine {
+        // Create a Foo from the FooBuilder, applying all settings in FooBuilder
+        // to Foo.
```
Comment (tdelabro) [inline src/vm/vm_core.rs:1115]: and this one

File: src/vm/vm_core.rs [hunk 4]
```diff
@@ -996,6 +996,138 @@ impl VirtualMachine {
...
+    pub(crate) memory: Memory,
+    pub(crate) accessed_addresses: Option<Vec<Relocatable>>,
+    pub(crate) trace: Option<Vec<TraceEntry>>,
+    pub(crate) current_step: usize,
+    skip_instruction_execution: bool,
+    run_finished: bool,
+}
+
+impl VirtualMachineBuilder {
+    pub fn new(trace_enabled: bool) -> VirtualMachineBuilder {
```
Comment (tdelabro) [inline src/vm/vm_core.rs:1013]: Why is trace enable treated differently than the other fields?
Couldn't it also have it setter method?

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: src/vm/vm_core.rs (hunk 4 above)
Comment (hurrikaanig) [reply src/vm/vm_core.rs:1013]: It also has one but I just reproduced VirtualMachine constructor behavior. Don't know if it's the right way of doing it or if it should be created empty.

**Comment by tdelabro** [comment]
also it's breaking the CI

File: src/vm/vm_core.rs [hunk 5]
```diff
@@ -986,6 +986,129 @@ impl VirtualMachine {
...
+    skip_instruction_execution: bool,
+    run_finished: bool,
+    #[cfg(feature = "hooks")]
+    pub(crate) hooks: crate::vm::hooks::Hooks,
+}
+
+impl Default for VirtualMachineBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
```
Comment (tdelabro) [inline src/vm/vm_core.rs:1006]: Perso si j'ai default et new qui font la même chose je préfère autant n'avoir que default.
C'est pas un red flag, mais c'est plus concis, respectueux des principe DRY

File: src/vm/vm_core.rs [hunk 6]
```diff
@@ -4015,4 +4138,40 @@ mod tests {
         let expected_traceback = vec![(Relocatable::from((1, 2)), Relocatable::from((0, 34)))];
         assert_eq!(vm.get_traceback_entries(), expected_traceback);
     }
+
+    #[test]
+    fn builder_test() {
+        let run_context = RunContext {
```
Comment (tdelabro) [inline src/vm/vm_core.rs:4144]: inline `RunContext { ... }` in the `VitualMachine { ... }` instanciation

File: src/vm/vm_core.rs [hunk 7]
```diff
@@ -4015,4 +4138,40 @@ mod tests {
...
+            pc: Relocatable::from((0, 0)),
+            ap: 0,
+            fp: 0,
+        };
+
+        let virtual_machine = VirtualMachine {
```
Comment (tdelabro) [inline src/vm/vm_core.rs:4150]: call it `expected_virtual_machine`

**Review by tdelabro — APPROVED** [review]

File: src/vm/vm_core.rs [hunk 8]
```diff
@@ -4015,4 +4134,38 @@ mod tests {
...
+            skip_instruction_execution: false,
+            segments: MemorySegmentManager::new(),
+            run_finished: true,
+            #[cfg(feature = "hooks")]
+            hooks: Default::default(),
+        };
+
+        let virtual_machine_from_builder: VirtualMachine = VirtualMachineBuilder::default()
+            .run_finished(true)
+            .current_step(12)
```
Comment (tdelabro) [inline src/vm/vm_core.rs:4160]: Add a call to each one of your setters function here

File: src/vm/vm_core.rs [hunk 9]
```diff
@@ -4015,4 +4134,90 @@ mod tests {
...
+            _constants: &HashMap<String, Felt>,
+        ) -> Result<(), VirtualMachineError> {
+            Ok(())
+        }
+
+        #[cfg(feature = "hooks")]
+        let _virtual_machine_from_builder: VirtualMachine = VirtualMachineBuilder::default()
+            .hooks(crate::vm::hooks::Hooks::new(
+                None,
+                Some(Arc::new(pre_step_hook)),
```
Comment (tdelabro) [inline src/vm/vm_core.rs:4218]: `Some(std::sync::Arc::new(pre_step_hook)),` will spare you the import of line 4200

File: src/vm/vm_core.rs [hunk 10]
```diff
@@ -4015,4 +4134,90 @@ mod tests {
...
+                public_memory_offsets: HashMap::new(),
+                memory: Memory::new(),
+            })
+            .skip_instruction_execution(true)
+            .trace(Some(vec![TraceEntry {
+                pc: Relocatable::from((0, 1)),
+                ap: Relocatable::from((0, 1)),
+                fp: Relocatable::from((0, 1)),
+            }]))
+            .build();
```
Comment (tdelabro) [inline src/vm/vm_core.rs:4166]: About the test of the hook.

Don't call `.build()` immediately. Assign the builder to a variable instead.

Then behind `#[cfg(feature = "hooks")]` call `hooks` on this variable.

Then call `build()` and store the vm_from_builder as you did.

Then do all your assert, then behind `#[cfg(feature = "hooks")]` add one more assert for the hooks

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: src/vm/vm_core.rs (hunk 10 above)
Comment (hurrikaanig) [reply src/vm/vm_core.rs:4166]: Smart !

**Review by tdelabro — APPROVED** [review]

**Review by fmoletta — APPROVED** [review]
Looks good! Thanks for the PR!

**Comment by tdelabro** [comment]
@Oppen would you like to be the second approver?

**Review by Jrigada — APPROVED** [review]
LGTM! Thanks

**Pull request merged**